### ISSUE TITLE
simplifying use of mathjax-node-sre

### ISSUE THREAD

**Issue by pkra** [pkra/mathjax-node-page#14]
As an author, I want my content to be as accessible as possible.

**Comment by pkra** [comment]
I don't see a natural way for doing this with npm. 

For the record, one just needs to change/add the dependency and change the `require('mathjax-node')` call in `lib/main.js`. Since SRE is a very large dependency, I'd rather not have it as a default dependency here.

**Comment by pkra** [comment]
If anyone has an idea how to provide alternative dependencies like this, please let me know.

**Comment by pkra** [comment]
Here's a simple idea.

```
try {
  const mathjax = require('mathjax-node-sre');
}
catch(e) {
  const mathjax = require('mathjax-node');
}
```

This way, if a project installs both mathjax-node-page and mathjax-node-sre, we switch automatically.

**Comment by pkra** [comment]
Actually more complicated, cf. https://gist.github.com/FGRibreau/3135914#gistcomment-579762 but seems to work. Not sure how to integrate it into the cli yet.

**Comment by PatrickMassot** [comment]
> For the record, one just needs to change/add the dependency and change the require('mathjax-node') call in lib/main.js.

Doing that seems to replace svg output with MathML output and do nothing else. What is the expected result?

**Comment by pkra** [comment]
Huh. That's weird. I have been extending mathjax-node-sre recently and something must have broken (and for obvious reasons there isn't a test for this yet :) )

**Comment by pkra** [comment]
Ah, I know what's happening. It's sort of a problem on both sides -- mathjax-node-sre always generated MathML (which it probably shouldn't) and mathjax-node-page blindly injects any output it finds instead of checking the configuration, with mathml coming last - thus overwriting previous svg output (which it probably shouldn't).

**Comment by PatrickMassot** [comment]
Thanks for 5d8f90c71167c13448. Now I can see the result. However I'm disappointed to see that neither lynx nor w3m display the title tag of inline svg whereas they both display the desc tag. Is there a reason to place the SRE output in title rather than desc?

Also it seems the original TeX source disappears completely from the output html. I think that, even with a very smart SRE, there will always be cases where users will want to read the source for disambiguation. I don't know where to put it and how to access it. But losing information seems unwise.

**Comment by pkra** [comment]
> Is there a reason to place the SRE output in title rather than desc?

Yes, better screenreader support.

> there will always be cases where users will want to read the source for disambiguation.

I understand the sentiment. I personally like spans behind the mathjax output which I hide but leave accessible to copy&paste (so you can copy a paragraph with TeX). 

Might be a nice optional feature.

**Comment by PatrickMassot** [comment]
>> Is there a reason to place the SRE output in title rather than desc?

> Yes, better screenreader support.

I'm not sure to understand what you call a screenreader here. Do you mean something like Orca? 

I know only one blind mathematician. He always says that software trying to make a graphical UI accessible (like Orca) are not usable. He browses the web using lynx (and a braille display driven by brltty) and reads the TeX source of maths papers when he wants to really understand. But he also listen to math papers (recorded by humans) when he wants to get a superficial understanding. 

So one could imagine something like lynx + festival reading a web page processed by mathjax-node-sre but with the possibility of stopping to read the source TeX at any time. Or simply using lynx, without speech output, but get the TeX source instead of formulas. This is still better than reading the full TeX source of a paper because you get hypertext links and less markup.

> I personally like spans behind the mathjax output which I hide but leave accessible to copy&paste.

Yes, this is also a very good reason to keep the source somewhere.

**Comment by pkra** [comment]
> I'm not sure to understand what you call a screenreader here.

Cf. https://en.wikipedia.org/wiki/Screen_reader. 

But I should have been more detailed: support across the accessibility tool chain which includes but is not limited to browsers populating the accessibility tree with suitable information, operating systems exposing that information in their accessibility APIs and screenreaders leveraging that information.

**Comment by pkra** [comment]
Fixed via #50

**Comment by PatrickMassot** [comment]
Could you elaborate on that "Fixed via #50"? What should I do to be able to run something like ``mjpage < in.html > out.html"`` and get math rendered as SVG together with SRE output somewhere (preferably in ``<desc>`` tag inside svg but if you insist on putting it only in ``<title>`` I can still run a postprocessor to fix it).

Also, did you also fixed the "keep source code somewhere" sub-issue?

**Comment by pkra** [comment]
> Could you elaborate on that "Fixed via #50"?

Please check the readme, e.g., https://github.com/pkra/mathjax-node-page/#mathjax-node-customization

> What should I do to be able to run something like mjpage < in.html > out.html"

You can't right now. Feel free to open a feature request for the CLI version, ideally with a PR.

> Also, did you also fixed the "keep source code somewhere" sub-issue?

No. If a solution is proposed others are ok with, I'll be happy to review it.
